File: 2023/01/day1.cpp
#include <iostream>
#include <vector>
#include <string>
#include <fstream>
#include <array>
#include <map>

#define INPUT_PATH "inputs/1.txt"

void readFileToVector(const char* filename, std::vector<std::string>& stringVector) {
	stringVector.clear();

	std::ifstream file(filename);
	std::string line;

	while (getline(file, line)) {
		stringVector.push_back(line);
	}
}

int firstAndLastDigitsCombined(std::string input) {
	int leftNumber = 0;
	int rightNumber = 0;

	for (int i = 0; i < input.size(); ++i) {
		char leftChar = input[i];

		if (leftChar >= '1' && leftChar <= '9') {
			leftNumber = leftChar - '0';
			break;
		}
	}

	for (int i = input.size() - 1; i >= 0; --i) {
		char rightChar = input[i];

		if (rightChar >= '1' && rightChar <= '9') {
			rightNumber = rightChar - '0';
			break;
		}
	}

	return (leftNumber * 10) + rightNumber;
}

int calculatePart1(std::vector<std::string>& input) {
	int total = 0;

	for (int i = 0; i < input.size(); ++i) {
		total += firstAndLastDigitsCombined(input[i]);
	}

	return total;
}

int calculatePart2Line(std::string input, std::array<std::string, 9>& digitStrings, std::map<std::string, int>& stringMap) {
	size_t firstIndex = input.size();
	size_t lastIndex = 0;
	int firstNumber = 0;
	int lastNumber = 0;

	for (int i = 0; i < input.size(); ++i) {
		char currentChar = input[i];

		if (currentChar >= '1' && currentChar <= '9') {
			firstNumber = currentChar - '0';
			firstIndex = i;
			break;
		}
	}

	for (int i = input.size() - 1; i >= 0; --i) {
		char currentChar = input[i];

		if (currentChar >= '1' && currentChar <= '9') {
			lastNumber = currentChar - '0';
			lastIndex = i;
			break;
		}
	}

	for (int i = 0; i < digitStrings.size(); ++i) {
		size_t firstStringIndex = input.find(digitStrings[i]);

		if (firstStringIndex != std::string::npos && firstStringIndex < firstIndex) {
			firstNumber = stringMap[digitStrings[i]];
			firstIndex = static_cast<int>(firstStringIndex);
		}

		size_t lastStringIndex = input.rfind(digitStrings[i]);

		if (lastStringIndex != std::string::npos && lastStringIndex > lastIndex) {
			lastNumber = stringMap[digitStrings[i]];
			lastIndex = static_cast<int>(lastStringIndex);
		}
	}

	return (firstNumber * 10) + lastNumber;
}

int calculatePart2(std::vector<std::string>& input) {
	int total = 0;
	std::array<std::string, 9> digitStrings = { "one", "two", "three", "four", "five", "six", "seven", "eight", "nine" };
	std::map<std::string, int> stringMap = {
		{"one", 1},
		{"two", 2},
		{"three", 3},
		{"four", 4},
		{"five", 5},
		{"six", 6},
		{"seven", 7},
		{"eight", 8},
		{"nine", 9},
	};

	for (int i = 0; i < input.size(); ++i) {
		total += calculatePart2Line(input[i], digitStrings, stringMap);
	}

	return total;
}

int main() {
	std::vector<std::string> input;
	readFileToVector(INPUT_PATH, input);

    int part1Solution = calculatePart1(input);

	std::cout << part1Solution << std::endl;

    int part2Solution = calculatePart2(input);

	std::cout << part2Solution << std::endl;

	return 0;
}

File: 2023/01/day1.rs
use twenty23::input;

use std::collections::HashMap;

const INPUT_PATH: &str = "inputs/1.txt";

const DIGIT_STRINGS: [&str; 9] = [
    "one", "two", "three", "four", "five", "six", "seven", "eight", "nine",
];

fn main() {
    let s = input::as_string(INPUT_PATH).unwrap();

    println!("{}", solve_part_1(&s));
    println!("{}", solve_part_2(&s));
}

fn solve_part_1(input: &str) -> u32 {
    let mut sum: u32 = 0;
    for line in input.lines() {
        let mut left_num: Option<u8> = None;
        let mut right_num: Option<u8> = None;

        for b in line.chars() {
            if b.is_digit(10) {
                left_num = Some(b as u8 - 48);
                break;
            }
        }

        for c in line.chars().rev() {
            if c.is_digit(10) {
                right_num = Some(c as u8 - 48);
                break;
            }
        }

        sum += 10 * left_num.unwrap() as u32 + right_num.unwrap() as u32;
    }

    sum
}

fn solve_part_2(input: &str) -> u32 {
    let map: HashMap<&str, u8> = HashMap::from([
        ("one", 1),
        ("two", 2),
        ("three", 3),
        ("four", 4),
        ("five", 5),
        ("six", 6),
        ("seven", 7),
        ("eight", 8),
        ("nine", 9),
    ]);

    let mut output: u32 = 0;

    for line in input.lines() {
        output += part_2_line(line, &map) as u32;
    }

    output
}

fn part_2_line(line: &str, map: &HashMap<&str, u8>) -> u8 {
    let mut first_index: usize = line.len();
    let mut last_index: usize = 0;

    let mut first_digit: u8 = 0;
    let mut last_digit: u8 = 0;

    let bytes = line.as_bytes();

    for i in 0..bytes.len() {
        let current = bytes[i];
        if current.is_ascii_digit() {
            first_digit = current - 48;
            first_index = i;
            break;
        }
    }

    for i in (0..bytes.len()).rev() {
        let current = bytes[i];
        if current.is_ascii_digit() {
            last_digit = current - 48;
            last_index = i;
            break;
        }
    }

    for i in 0..DIGIT_STRINGS.len() {
        if let Some(first_string_index) = line.find(DIGIT_STRINGS[i]) {
            if first_string_index < first_index {
                first_digit = *map.get(DIGIT_STRINGS[i]).unwrap();
                first_index = first_string_index;
            }
        }
        if let Some(last_string_index) = line.rfind(DIGIT_STRINGS[i]) {
            if last_string_index > last_index {
                last_digit = *map.get(DIGIT_STRINGS[i]).unwrap();
                last_index = last_string_index;
            }
        }
    }

    10 * first_digit + last_digit
}
